### ISSUE TITLE
Build of gcc 5.3.0 fails on Arch Linux

### ISSUE THREAD

**Issue by szaghi** [sourceryinstitute/opencoarrays#188]
Dear all,

I am reporting this failing on Arch Linux, but I am almost sure that this issue is due to Arch Linux and not to your great installer. Anyhow, I know you may want to have any feedback.

I tried to build and install gcc 5.3.0 with the following command

```bash
stefano@zaghi(10:46 AM Fri May 20) on master
→ ./build.sh -p gcc -c -f -j 10 -I 5.3.0 -i /opt/arch/gcc/5.3.0
```

Presently, Arch Linux has gcc 6.1.0 as default and trying to install also the 5.x trunk is painful. As a matter of fact, the manually building from source of gcc 5.3.0 fails in the same way the building by means of your script, thus I am quite sure that this problem is due to Arch Linux. For the sake of completeness I attached the log file of failing build.

[gcc-v5.3.0-building-fails.log.zip](https://github.com/sourceryinstitute/opencoarrays/files/274218/gcc-v5.3.0-building-fails.log.zip)

I report this in case you like to advice other Archers that this is an issue of Arch Linux and not of the installer script.

Close this as soon you want.

My best regards.

Stefano

**Comment by zbeekman** [comment]
I haven't dug into the specifics, but I know multi-threaded GCC builds often fail. When installing OpenCoarrays directly, the script uses a single threaded GCC build. Could you try removing the `-j 10` flag?

**Comment by szaghi** [comment]
@zbeekman Yes, I can surely test this. Indeed, I am quite sure that the last time I have installed gcc (gcc 6 before it was a default for Arch) I not changed my default make settings, but my memory could be wrong, and you have always the right answers :smile: I do a check right now, stay tuned!

**Comment by szaghi** [comment]
:cry: no way...

I tried with both your build script and manually disabling multi-thread compilation, but it is still failing

[gcc-v5.3.0-building-fails.log.zip](https://github.com/sourceryinstitute/opencoarrays/files/274781/gcc-v5.3.0-building-fails.log.zip)

Anyway, do not care about this: I am almost sure that this is not an issue related to your installer, but due to Arch Linux, as a matter of facts also the *manual way* fails similarly.

Thank you Zaak for your help.

**Comment by rouson** [comment]
I also haven't looked into this in detail and won't be able to for a few hours at least.  

I don't understand the empty `-c -f -j` flags.  Each of those flags requires an argument, e.g.,
```
./build.sh -p gcc -c /opt/local/bin/gcc -j /opt/local/bin/gfortran -C /opt/local/bin/g++ -I 5.3.0 \
-i /opt/arch/gcc/5.3.0
```

Please try the above.

**Comment by szaghi** [comment]
@rouson Right, sorry for my too fast read of build help, I am re-run without them (this should use the defaults, right?)

```bash
./build.sh -p gcc -I 5.3.0 -i /opt/arch/gcc/5.3.0
```

**Comment by rouson** [comment]
Yes. That should use the defaults.  How did it go?

**Comment by rouson** [comment]
Also, I just realized a couple of things that I should clean up in a patch release:
1. I'm thinking the documentation will now recommend using ./install.sh for all purposes. Although it should work to call ./build.sh directly, I don't think there are any specific reasons to do so at the moment. install.sh invokes build.sh but has additional capabilities so the recommended command is
```
./install.sh -p gcc -I 5.3.0 -i /opt/arch/gcc/5.3.0
```
2. The documentation and error-detection should make it clear that which flags require an argument (such as is the case for all flags we've discussed).  This is a very easy fix that we should be able to get post in a bug fix release (1.6.1) next week. In fact, the -f, -c, -C have no default values defined, but on quick review, it appears the build.sh uses the values of the FC, CC, and CXX environment variables if they are set.   If you have any preferences or opinions on whether FC should override -f or vice versa, please let us know.

**Comment by rouson** [comment]
@szaghi, I'm not sure if this is related, but I just created Issue #189 to demonstrate that we need better detection of missing arguments.  I'll ask @zbeekman to look into this.

**Comment by szaghi** [comment]
@rouson @zbeekman I am sorry for my delay. I cannot say if the last tentative is worked: I had to left the office yesterday after the launch of it (it happens that sometimes my wife like to see me one hour per day, she has strange preferences :smile: ).

Later today I will try to ssh my workdtation and check the log, but with only my smartphone I cannot promise the success.

Thank very much for your help.

**Comment by rouson** [comment]
No rush.  Thanks for the feedback.

**Comment by szaghi** [comment]
@rouson @zbeekman just checked, still failing, the exit seems the same of above, but I cannot diff the log with the older with my smartphone. Monday I can upload the log if you want.

Have a good weekend, cheers.

**Comment by zbeekman** [comment]
Yes seeing the log may be helpful.

**Comment by szaghi** [comment]
@zbeekman 

here is the log:

[gcc-v5.3.0-building-fails.log.zip](https://github.com/sourceryinstitute/opencoarrays/files/277051/gcc-v5.3.0-building-fails.log.zip)

The test was 

```bash
./build.sh -p gcc -I 5.3.0 -i /opt/arch/gcc/5.3.0 > gcc-v5.3.0-building-fails.log 2>&1
```

**Comment by szaghi** [comment]
Anyhow, I really think the problem is due to Arch Linux update: it being a rolling release it could happen that updating the system makes very complex the support of *older* programs that before the update work well. As a matter of facts, I am not able to install gcc 5.3 manually, whereas I was able to do it before the update.

**Comment by rouson** [comment]
Thanks for this additional feedback. Given that we don't have much idea about what to do and you suspect the problem is not on our side, I'll close this issue.  If you disagree, please feel free to reopen it.  We appreciate your continued interest in OpenCoarrays and value the issue reports.

**Comment by zbeekman** [comment]
Hi @szaghi,
It seems that this issue is caused when trying to build an older GCC with GCC 6. It appears they have patched gcc-5 branch to allow it to be built with GCC 6, however, if you want to build 5.3 it seems like you will have to use an older compiler. If you have an older GCC still installed on your system, `install.sh` has the `-c`, `-C`, and `-f` flags to let you specify which compiler to use to perform the build. This might let you build 5.3 with an older compiler (4.8 for example...).

Here are the details on the bug tracker: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=69959

**Comment by szaghi** [comment]
@zbeekman you are better than Sherlok Holmes. I guessed somenthing along this line (I was betting it was due to some newer version of dependencies, but not to gcc itself...), but I will never find the bugzilla issue, great Zaak!

Feel free to close this (tomorrow I will try to find solution, but I am pretty sure that is very difficult with my current Arch).

**Comment by zbeekman** [comment]
Yup, issue is closed already... as I mentioned if you have another, older version of GCC hanging around somewhere on your system, the `install.sh` script can be told to use a different compiler by specifying the full path to the `-c`, `-C` and `-f` flags....

**Comment by szaghi** [comment]
ops... I have to improve my github-smartphone experience. Unfortunately I have only gcc 6.x (6.0, 6.1 manually installed when Arch was on 5.3 and 6.1 now in the system)...

**Comment by zbeekman** [comment]
ok, well in that case, I don't know if you'll be able to get 5.3 installed... you might be able to install the gcc-5 branch using the script--it sounded like they committed changes to enable building with 6.x ... I'm not sure it's currently possible to build the gcc-5 branch with the script, but it looks like Damian may be adding this capability.

I'm surprised theres no older GCC lying around somewhere like /usr/bin or /bin etc.

**Comment by szaghi** [comment]
When Arch updates the system, all the system must be *self-referenced*... partial updates are not supported for rolling-release distro, so when gcc is updated pacman updates just /usr/bin exes... the only hope is that some Archer like me provides an aur package build for re-add gcc 5.x trunk.

**Comment by zbeekman** [comment]
well that sounds like a bit of a pain... but at least you don't have to wait too long for packages I would guess

**Comment by szaghi** [comment]
@zbeekman 

### The happy end

Some Archers have had *pity* of me: to whom may concerns, now there is an AUR package build to compile gcc 5.3.0 alongside the official gcc 6.1.1

https://aur.archlinux.org/packages/gcc5/

For the sake of completeness, I report in the following the build script that successfully built gcc 5.3.0 on my Arch Linux.

The package build script is quite verbose (the configure invocation has many flags that I did not use)

```bash
# $Id$
# Maintainer: Sebastiaan Lokhorst <sebastiaanlokhorst@gmail.com>
# Contributor: Ruben Van Boxem <vanboxem.ruben@gmail.com>
# Contributor: Allan McRae <allan@archlinux.org>

pkgname=('gcc5')
pkgver=5.3.0
_pkgver=5
_islver=0.15
pkgrel=6
_snapshot=5-20160503
pkgdesc="The GNU Compiler Collection"
arch=('i686' 'x86_64')
license=('GPL' 'LGPL' 'FDL' 'custom')
url="http://gcc.gnu.org"
makedepends=('binutils>=2.26' 'libmpc')
options=('!emptydirs')
source=(#ftp://gcc.gnu.org/pub/gcc/releases/gcc-${pkgver}/gcc-${pkgver}.tar.bz2
        ftp://gcc.gnu.org/pub/gcc/snapshots/${_snapshot}/gcc-${_snapshot}.tar.bz2
        http://isl.gforge.inria.fr/isl-${_islver}.tar.bz2)
md5sums=('907f3b860ebbaf33b3ccd3c15dddc9b1'
         '8428efbbc6f6e2810ce5c1ba73ecf98c')

#_basedir=gcc-${pkgver}
_basedir=gcc-${_snapshot}

prepare() {
  cd ${srcdir}/${_basedir}

  # link isl for in-tree build
  ln -s ../isl-${_islver} isl

  # Do not run fixincludes
  sed -i 's@\./fixinc\.sh@-c true@' gcc/Makefile.in

  # Arch Linux installs x86_64 libraries /lib
  [[ $CARCH == "x86_64" ]] && sed -i '/m64=/s/lib64/lib/' gcc/config/i386/t-linux64

  echo ${pkgver} > gcc/BASE-VER

  # hack! - some configure tests for header files using "$CPP $CPPFLAGS"
  sed -i "/ac_cpp=/s/\$CPPFLAGS/\$CPPFLAGS -O2/" {libiberty,gcc}/configure

  mkdir ${srcdir}/gcc-build
}

build() {
  cd ${srcdir}/gcc-build

  # using -pipe causes spurious test-suite failures
  # http://gcc.gnu.org/bugzilla/show_bug.cgi?id=48565
  CFLAGS=${CFLAGS/-pipe/}
  CXXFLAGS=${CXXFLAGS/-pipe/}

  ${srcdir}/${_basedir}/configure --prefix=/usr \
      --build=${CHOST} \
      --libdir=/usr/lib --libexecdir=/usr/lib \
      --mandir=/usr/share/man --infodir=/usr/share/info \
      --with-bugurl=https://bugs.archlinux.org/ \
      --enable-languages=c,c++,fortran,go,lto,objc,obj-c++ \
      --enable-shared --enable-threads=posix --enable-libmpx \
      --with-system-zlib --with-isl --enable-__cxa_atexit \
      --disable-libunwind-exceptions --enable-clocale=gnu \
      --disable-libstdcxx-pch --disable-libssp \
      --enable-gnu-unique-object --enable-linker-build-id \
      --enable-lto --enable-plugin \
      --with-linker-hash-style=gnu --enable-gnu-indirect-function \
      --disable-multilib --disable-werror \
      --enable-checking=release \
      --program-suffix=-${_pkgver} \
      --enable-version-specific-runtime-libs
      #--enable-install-libiberty

  export LD_PRELOAD=/usr/lib/libstdc++.so
  make
}

package() {
  cd ${srcdir}/gcc-build

  make -j1 DESTDIR=${pkgdir} install
  
  # Lazy way of dealing with conflicting files...
  rm -rf ${pkgdir}/usr/share/{info,locale,man}

  # Move potentially conflicting stuff to version specific subdirectory
  mv ${pkgdir}/usr/lib/lib* ${pkgdir}/usr/lib/gcc/${CHOST}/${pkgver}/
  mv ${pkgdir}/usr/lib/gcc/${CHOST}/lib/lib* ${pkgdir}/usr/lib/gcc/${CHOST}/${pkgver}/
  
  # Install Runtime Library Exception
  install -Dm644 ${srcdir}/${_basedir}/COPYING.RUNTIME \
    ${pkgdir}/usr/share/licenses/$pkgname/RUNTIME.LIBRARY.EXCEPTION
}
```

**Comment by zbeekman** [comment]
I'm glad you got it working for arch linux... Another approach I would have recommended would be to install via http://linuxbrew.sh

**Comment by rouson** [comment]
Shall we incorporate this script into our installation scripts? Obviously we won't attempt anything close comprehensive coverage of all distros and all compiler combinations, but since @szaghi was nice enough to share this script with us, we might as well innoculate our build system against this scenario.  Thoughts?

**Comment by zbeekman** [comment]
My preference would be no. The installation machinery is already quite sophisticated and this is a corner case that seems to be well handled through the usual means.... also, it sounds like arch linux moves quite fast and has no notion of package versions. This script is only to build older versions of GCC with GCC 6.1 on arch linux a very specialized case.

**Comment by zbeekman** [comment]
@rouson unless I am misunderstanding you, which may be possible. If you mean, attempt to adapt our current GCC installation process to be able to use some of the configure options etc. so it might work on arch linux (without breaking on ubuntu and OS X) then I'd be willing to try, but that seems like a lower priority.

**Comment by rouson** [comment]
I agree with your first response.  We can skip it. 

Regarding the second response, I was just thinking of a simple conditional to check for the specific situation that started this thread and then a single line invocation of something very close to the script that @szaghi pasted into this thread.  

I'm making an analogy to an immune system in which the conditional and submitted script play the role of an antibody should our build system encounter this situation again.  The benefit might not be huge if this scenario is unlikely to occur again, but the cost is also small, given that the script has been written for us. 

Then again, I'm no biologist. :)

**Comment by szaghi** [comment]
@zbeekman @rouson 

My dear, Arch Linux is a very *corner* distro, similar to Gentoo, I feel that you can forget about it. My intention was just to give you my feedback. 

For the sake of completeness, I would like to give you some details of Arch philosopy. It is a rolling release, so it updates *continuosly*. It has the notion of packages version, but it is designed to work (indeed very well) with *that version* for *that update*, e.g. the current update is designed to work with gcc 6.1.1. In general, installing older version is very simple and do not affect the system: this is the first time I have had such a pain (as Zaak point out because of gcc 5.x has problems to be built with gcc 6.x), but at the end the Archers community is very fast to solve the problems. Indeed, this approach works very well, you have a very update distro while the system remains stable. There are cons too: partial updates are not supported. You can decide if update your system or not (I do almost every day), but if you update it you update all the system, not single packages.

Closed this Arch divagation, here are some of my poor thougths for what can be valued: your installer is great, but already sophisticated/complex enough, trying to add a so comprhensive support for very corner case like mine can make installer *complicated* without adding a real value. Very special case are better solved with special solution. I am aligned with Zaak. If Damian still would help donkey like me a lightweigth approach could be to add some *descriptive informations of the possible issues* to the script. For example, if another Archer fails into my case (use the installer to install gcc 5.3 in the current Arch) the installer (after the failing) could recognize that the OS is Arch and prompt to the user 

> dear archer your OS has known issues for installing older gcc, try to use the package build [gcc5](link-to-the-aur-package-build)

In other words, let Archers to help Archers with a minimum impact on the installer complexity.

This approach has its own cons: you should take care to update the list of known issues. For Arch I can help you, but for Gentoo, Slackware, OpenSuse...? 

My feel is that the installer is **already great** and you do not need to make it **great again** due to my very corner case.

My best regards.

**Comment by rouson** [comment]
@zbeekman and @szaghi, many thanks to both of you for your input. :clap:  It sounds like we can move on to solving other problems.
